src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3097,24 +3104,92 @@ void addObjectEquals(MethodSymbol thisEquals)
 
             void addHashCode()
             {
-                var hashCode = new SynthesizedRecordGetHashCode(this);
+                var hashCode = new SynthesizedRecordGetHashCode(this, memberOffset: members.Count);
                 if (!memberSignatures.ContainsKey(hashCode))
                 {
                     // https://github.com/dotnet/roslyn/issues/44617: Don't add if the overridden method is sealed
                     members.Add(hashCode);
                 }
             }
 
-            MethodSymbol addThisEquals()
+            static PropertySymbol? getInheritedEqualityContract(NamedTypeSymbol type)
+            {
+                while ((type = type.BaseTypeNoUseSiteDiagnostics) is object)
+                {
+                    var members = type.GetMembers(SynthesizedRecordEqualityContractProperty.PropertyName);
+                    // https://github.com/dotnet/roslyn/issues/44903: Check explicit member has expected signature.
+                    if (members.FirstOrDefault(m => m is PropertySymbol property && property.ParameterCount == 0) is PropertySymbol property)
+                    {
+                        return property;
+                    }
+                }
+                return null;
+            }
+
+            PropertySymbol addEqualityContract()
+            {
+                var property = new SynthesizedRecordEqualityContractProperty(this, isOverride: getInheritedEqualityContract(this) is object);
+                // https://github.com/dotnet/roslyn/issues/44903: Check explicit member has expected signature.
+                if (!memberSignatures.ContainsKey(property))
```

Comment:
not blocking this PR: should we check both whether the property and the get accessor are already in the list of members? Could we have the get accessor defined without property?